### PR TITLE
Many Changes

### LINKED ISSUE #12: 0.7 Cleanup
with 0.7 done, it's time to clean up the code. Remove scafolding and make sure all the documentation is current, as well as making sure there are tests for each of the issues in this milestone.

### LINKED ISSUE #80: No way of configuring Scope in connections.py
Connections.oauth2 has no way of changing the scope for the token, as it is hardcoded.

I wanted to read my Calendars, but since Calendars.Read was not requested I was not allowed even if I gave the app those permissions. So I had to hardcode that in the source as well.

### PR DESCRIPTION
**Pull request by roycem90**
**Enable "Github Pages" with option "master branch /docs folder" under settings**

- Message is moved to message.py (but available to import directly from O365)
- FluentMessage is added - A Fluent API wrapper for Message
- Many methods in Message are deprecated to support python convention instead of camelCase
- Properties added in Message to support quick access (Closes #12 )
- Bug #114 is fixed with update of Message class to use Connection api
- Scopes can now be set using Connection.scopes (Closes #80)
- Added Documentation using HTML (removed related content from ReadMe.md)
- Update connection to work on python 2 as per #79 
- Fixes in MicroDict to handle case differences between v1.0 (PascalCase) and 2.0 (camelCase) with code using the 2.0 case

Guess thats all for now :)

### CODE REVIEW

**Comment by roycem90** [comment]
@Narcolapser - not doing a merge as it requires you to update the settings

**Comment by roycem90** [comment]
you can have a look at documentation at https://o365.github.io/ - for how it looks and also the changes are clearly visible  :100: 
that is temporary, it has to be manually copied from sphinx build and commited 

using the docs/ folder with github pages gives a little longer url but easily maitainable

**Comment by roycem90** [comment]
yaa, sounds reasonable i guess u can decide on how much to keep on the read me,

as per some libraries i checked most have installation, contributing and other instructions and on ReadMe.md and "Getting started" goes into the doc link, so i did the same way as it felt redundant to have the same examples here and there..

you can pickup the required parts and copy them over back to readme. (you can look as the latest docs by cloning the branch and open docs/index.html)

**Comment by roycem90** [comment]
reverted the readMe file and just added the URL to the new docs alone, 
added reply and reply_all methods as asked in #132

**Comment by roycem90** [comment]
we have to migrate our calendar stuff alto to use Connection (Nov 1st will drop the access to basic auth), is anyone working on it?

**Comment by Narcolapser** [comment]
@Janscas has gotten most of it up and running, I just need to get around to merging his code. He seems to have gone his own way? I'm not sure, maybe he can come here and correct me on that. But I think I have time this afternoon to fiddle with this so I'll give that a try.

Also, I've got tests occurring but all the tests will need some level of rewriting. Once I get that handled I'll see about getting automatic publishing in place.

**Comment by Narcolapser** [comment]
ooooooih. I really wish he would have worked with us on this instead of running of and creating the whole thing on his own. There will be no merging his code. Either we just supplant everything we've got with his code or there is two libraries, python-o365 and py-o365. I'd rather do the former, but I don't want to unless he's willing to work with us. How about it @janscas did you want there to be one library or do you want to run your own fork?

**Comment by janscas** [comment]
### TL;DR: Yes I've been working on the project and it's almost complete and yes, as stated on previous conversations I continue to think there must be only one library. 
___

Hi there!

Yes I've been working on the project on my own.

First of all I didn't saw any further movement on O365 to rebuild the library for real and second and most important I had to keep moving on my own because I needed the functionality ready for my current job.
That's why I kept developing the library mainly but also because I don't like some design choices on the O365 library such as:

- Connection singleton (you can't connect with two accounts on the same script).
- Having FluentMessage and Message (why not choose the better one).
- Little interoperativity between components (a Contact has to be able to be used and understood between a Message an Event or a Drive when sharing for example).
- Hardcoding url paths on Class Attributes: If you hardcode '/__me__/contacts/' how can I change it to '/__users/user_id__/contacts? 
- Etc.

At these point (7 months have passed) I have everything almost done except for sharepoint and some other few things (I have already dropped support for the Office 365 API v1 altogether because of the basic auth stuff and also the properties naming mess). There are however a lot of things that need to be refactored and adjusted. I'm pretty happy with the current state of the library and the functionality it has.

But as I said on other posts, I really feel this work must be merged together. So as before I'm willing to start the conversation, merge parts of the code between the projects and work together to achieve a solid and 100% functional and well maintained library. My only concerns are __a)__ don't loose functionality over what's already done __b)__ use pythonic an pep8 compliant code __c)__ library must be written *for humans*

**Comment by roycem90** [comment]
@jancas
1. Connection Singleton - I added it as singleton to avoid passing it to every request, and i didn't see anyone saying about using another account at the same time so never made any changes to that, though it can be fixed pretty easily (without making major changes)
2. FluentMessage and Message - Fluent API's are generally good to build stuff like build a message or request etc..., but reading a bunch of fields like subject, body etc.. is not great with fluent interface, hence i added a simple Message to be able to use it normally and wrapped a fluent interface around it
3. Haven't given a look at Event/Drive/Contact, so can't comment on it
4. i think we have url's for both *me* and */user/user_id* added where applicable in the class url's

and for the merging i don't think it would that easy given most of the api/code is completely re-written, we may to do them in part by part basis or something like that

**Comment by Narcolapser** [comment]
Good to hear!

On the whole I am in agreement here. Every flaw that was pointed out was a known issue that exists entirely because I just never took the time to update it. About the time I would have been doing that I lost my old O365 account and couldn't do anything, it was about a year later that I got a new O365 (both old and new from are my employers, I don't have a personal one) and I never really got back into the swing of things. So I am glad to see you've addressed them. And thank you for explaining your absence, so to speak.

As for merging I see the following options:

1. Supplant everything in one fell swoop.
2. Bring over everything that doesn't have a name space conflict, mark everything else as deprecated, push a release, wait a week or two and migrate the rest and push a fully merged release.
3. Gut the current library and rebuilt it's interface running on top of the new library. Mark the interface as deprecated and remove it in the next release.

As crazy as it is I lean towards option 1. It is just nice and simple. When we do I'd say it will be version 1.0, implying that the API has changed. Thoughts? Other options I didn't think of?

**Comment by roycem90** [comment]
one limitation is pyo365 requires python 3.4+

**Comment by Narcolapser** [comment]
That is a good point. I always like to maintain as much compatibility as possible. I will admit that I still default to python 2.7 almost entirely because it is the default install on my Ubuntu machines. But there is a part of me that is fine with dropping legacy support. I wonder how much it would take to support 2.7...

**Comment by roycem90** [comment]
depends on what all he used, can we pull his repo onto one of the branch here??

**Comment by janscas** [comment]
@roycem90 I understand the decisions behind the Connection singleton, just saying that it'has it's limitations. Also note that the hardcoded urls are almost all set to '/me' and sometimes to '/users/user_id' but this is not enough... what if I want to access a Office365 Group? where can I set the resource to 'group/group_id'?

@Narcolapser It's ok to pull all. But If that's the case I will like you and @roycem90 to see if it fits, analyse what parts can be optimised or rewrite, etc.

@roycem90 Yes, it's only python 3 and only >=3.4 version.Python 2.7 it's going to off in 1 year and 2 months ([pythonclock](https://pythonclock.org/)) I don't even care about python 2 anymore.
I could tell you it's not compatible with 2.7 in a lot of ways. Dict iterations, Enums, pathlib, etc.. Making it 2.7 compatible it's gonna be a hard work that I'm not willing to do. You can of course do it.

**Why don't you both take a look at my code, see what you like or not, ask some questions, and see if you finally want to use it and merge it ... all, something or not at all?**

**Comment by janscas** [comment]
See also [python3statement.org](https://python3statement.org/). Scroll down and see.

A lot of popular libraries are dropping support for python 2.7 soon. Pandas for example will do it on January 1st 2019.

**Comment by Narcolapser** [comment]
Me being facetious: When I put those hard coded urls in place now 4 years ago, they were the only urls. :p

I think the jump to the new code base would be a good time to drop python2 support. I hadn't seen the pythonclock, that makes me smile. I've poked around with your code a little, I'll do some more reading today.

**Comment by roycem90** [comment]
lol never knew about the python2 support drop

anyway i checked most of the code on project and except the keyword only arguments, everything else can be made workable on python 2..

but given support drop i think we can go ahead with the merge, lets put a note that people interested in using python 2.7 should use the old version (i mean the existing one, with it as bug fixes only mode)

what say?

**Comment by roycem90** [comment]
and one more thing the package name has to be renamed from pyo365 back to O365

**Comment by janscas** [comment]
> Check this https://devguide.python.org/#status-of-python-branches
> even python 3.4 is being killed by March, 2019

But that Is not important because

**Comment by roycem90** [comment]
@janscas why did you close this, this was bunch of fixes for the existing version

**Comment by janscas** [comment]
@roycem didn’t know I close it. My mistake... sry

**Comment by roycem90** [comment]
anyway as we are merging the rewrite i will close this and rename the branch to something like python2.7 or old_o365 kind of :)
